core/Makefile

```diff
@@ -112,20 +137,28 @@ LDFLAGS += -rdynamic -L$(DPDK_LIB_DIR) -Wl,-rpath=$(DPDK_LIB_DIR) -pthread
 ifdef BESS_LINK_DYNAMIC
     LIBS_ALL_SHARED = -Wl,-call_shared
     LIBS_DL_SHARED =
+    ifeq ($(HAS_PKG_CONFIG), yes)
+      PKG_LIBS := $(shell $(PKG_CONFIG) --libs $(PKG_CONFIG_DEPS))
+    else
+      PKG_LIBS := $(NO_PKG_CONFIG_LIBS)
+    endif
 else # Used static libraries
     LIBS_ALL_SHARED =
     LIBS_DL_SHARED = -Wl,-call_shared
-    LIBS_LZMA = -llzma
     LDFLAGS += -static-libstdc++
+    ifeq ($(HAS_PKG_CONFIG), yes)
+      PKG_LIBS := $(shell $(PKG_CONFIG) --static --libs $(PKG_CONFIG_DEPS))
+    else
+      PKG_LIBS := $(NO_PKG_CONFIG_LIBS) $(NO_PKG_CONFIG_LIBS_INDIRECT)
+    endif
 endif
 
-LIBS += -Wl,-non_shared \
+LIBS += $(ALWAYS_DYN_LIBS) -Wl,-non_shared \
 	-Wl,--whole-archive -l$(DPDK_LIB) -Wl,--no-whole-archive \
 	$(LIBS_ALL_SHARED) \
-	-lglog -lgflags -lprotobuf -lgrpc++ -lgrpc \
-	-lssl -lcrypto -lunwind $(LIBS_LZMA) -lpcap -lz \
+	$(PKG_LIBS) $(ALWAYS_LIBS) \
 	$(LIBS_DL_SHARED) \
-	-ldl
+	$(ALWAYS_DYN_LIBS)
```

Comment:
Just checking: Adding `$(ALWAYS_DYN_LIBS)` twice was intended?

Reply:
It was intended.

DPDK depends on `-ldl`, so `-ldl` must be specified after DPDK.

But then again, we want to dynamically link with `-ldl`.  If one of the other libs depend on `-ldl` it will be listed in the static section.  We must make sure that we include it at the beginning of the cmdline, in the dynamic section